File: msf/catalog_delta.go
package msf

import (
	"encoding/json"
	"fmt"
	"slices"
)

// CatalogDelta represents a delta update for an existing MSF catalog.
//
// A delta catalog never carries a complete track list. Instead it contains one
// or more add/remove/clone operations plus optional metadata updates. The JSON
// form (draft-ietf-moq-msf-01) is a "deltaUpdate" array of operation objects,
// each shaped as {"op": "add"|"remove"|"clone", "tracks": [...]}.
//
// The in-memory model keeps three grouped slices (AddTracks/RemoveTracks/
// CloneTracks) and records the first-seen block order of each op kind so the
// wire array round-trips. Interleaved same-type ops (e.g. add,remove,add) are
// merged into a single group; the resulting catalog is identical in practice.
type CatalogDelta struct {
	// DefaultNamespace, if set, replaces the base catalog namespace used for
	// resolving tracks whose namespace field is omitted.
	DefaultNamespace string `json:"-"`
	// GeneratedAt, if non-nil, records the delta's generatedAt timestamp.
	GeneratedAt *int64 `json:"-"`
	// IsComplete marks the updated catalog as complete.
	IsComplete bool `json:"-"`

	// AddTracks lists new track entries to append to the base catalog.
	AddTracks []Track `json:"-"`
	// RemoveTracks identifies tracks to remove from the base catalog.
	RemoveTracks []TrackRef `json:"-"`
	// CloneTracks lists clone operations derived from an existing parent track.
	CloneTracks []TrackClone `json:"-"`

	// ExtraFields stores unknown JSON properties for round-tripping.
	ExtraFields map[string]json.RawMessage `json:"-"`

	deltaOpOrder []deltaOperationKind
}

// TrackRef identifies a track by namespace and name.
//
// It is used where only the identity of a track is required, such as
// removeTracks entries inside a catalog delta.
type TrackRef struct {
	Namespace string `json:"-"`
	Name      string `json:"-"`

	ExtraFields map[string]json.RawMessage `json:"-"`
}

// TrackClone describes a clone operation in a catalog delta.
//
// The embedded Track provides the override values to apply to the cloned track.
// ParentName identifies the source track from which values are inherited and
// ParentNamespace (optional) identifies the namespace of that source track.
type TrackClone struct {
	Track

	ParentName      string `json:"-"`
	ParentNamespace string `json:"-"`
}

// Clone returns a deep copy of the delta catalog.
func (d CatalogDelta) Clone() CatalogDelta {
	clone := d
	clone.AddTracks = cloneTracks(d.AddTracks)
	clone.RemoveTracks = cloneTrackRefs(d.RemoveTracks)
	clone.CloneTracks = cloneTrackClones(d.CloneTracks)
	clone.ExtraFields = cloneRawMessages(d.ExtraFields)
	clone.deltaOpOrder = slices.Clone(d.deltaOpOrder)
	return clone
}

// Validate checks whether the delta satisfies the package's MSF draft-01 rules.
func (d CatalogDelta) Validate() error {
	var problems []string

	if len(d.AddTracks) == 0 && len(d.RemoveTracks) == 0 && len(d.CloneTracks) == 0 {
		problems = append(problems, "delta catalog must contain addTracks, removeTracks, or cloneTracks")
	}
	for i := range d.AddTracks {
		if errs := d.AddTracks[i].validate(""); len(errs) > 0 {
			prefix := "addTracks[" + itoa(i) + "]: "
			for _, err := range errs {
				problems = append(problems, prefix+err)
			}
		}
	}
	for i := range d.RemoveTracks {
		if errs := d.RemoveTracks[i].Validate(""); len(errs) > 0 {
			prefix := "removeTracks[" + itoa(i) + "]: "
			for _, err := range errs {
				problems = append(problems, prefix+err)
			}
		}
	}
	for i := range d.CloneTracks {
		if errs := d.CloneTracks[i].Validate(""); len(errs) > 0 {
			prefix := "cloneTracks[" + itoa(i) + "]: "
			for _, err := range errs {
				problems = append(problems, prefix+err)
			}
		}
	}

	return newValidationError(problems)
}

// ParseCatalogDelta decodes an MSF catalog delta from JSON bytes.
func ParseCatalogDelta(data []byte) (CatalogDelta, error) {
	var delta CatalogDelta
	if err := json.Unmarshal(data, &delta); err != nil {
		return CatalogDelta{}, err
	}
	return delta, nil
}

// ParseCatalogDeltaString is like ParseCatalogDelta but accepts a string.
func ParseCatalogDeltaString(s string) (CatalogDelta, error) {
	return ParseCatalogDelta([]byte(s))
}

// operationOrder returns the declared delta operation order, preserving JSON order when known.
func (d CatalogDelta) operationOrder() []deltaOperationKind {
	if len(d.deltaOpOrder) > 0 {
		return slices.Clone(d.deltaOpOrder)
	}

	order := make([]deltaOperationKind, 0, 3)
	if len(d.AddTracks) > 0 {
		order = append(order, deltaOperationAdd)
	}
	if len(d.RemoveTracks) > 0 {
		order = append(order, deltaOperationRemove)
	}
	if len(d.CloneTracks) > 0 {
		order = append(order, deltaOperationClone)
	}
	return order
}

var (
	_ json.Marshaler   = CatalogDelta{}
	_ json.Unmarshaler = (*CatalogDelta)(nil)
	_ json.Marshaler   = TrackRef{}
	_ json.Unmarshaler = (*TrackRef)(nil)
	_ json.Marshaler   = TrackClone{}
	_ json.Unmarshaler = (*TrackClone)(nil)
)

// MarshalJSON encodes the delta in the draft-01 JSON form: a "deltaUpdate"
// array of {"op", "tracks"} objects in declared operation order.
func (d CatalogDelta) MarshalJSON() ([]byte, error) {
	obj := make(map[string]any, len(d.ExtraFields)+2)
	for key, raw := range d.ExtraFields {
		obj[key] = cloneRawMessage(raw)
	}
	ops := make([]map[string]any, 0, 3)
	for _, op := range d.operationOrder() {
		entry := map[string]any{"op": string(op)}
		switch op {
		case deltaOperationAdd:
			entry["tracks"] = d.AddTracks
		case deltaOperationRemove:
			entry["tracks"] = d.RemoveTracks
		case deltaOperationClone:
			entry["tracks"] = d.CloneTracks
		}
		ops = append(ops, entry)
	}
	obj["deltaUpdate"] = ops
	if d.GeneratedAt != nil {
		obj["generatedAt"] = *d.GeneratedAt
	}
	if d.IsComplete {
		obj["isComplete"] = true
	}
	return json.Marshal(obj)
}

// UnmarshalJSON decodes a delta catalog and rejects independent-catalog fields.
func (d *CatalogDelta) UnmarshalJSON(data []byte) error {
	*d = CatalogDelta{}
	d.ExtraFields = make(map[string]json.RawMessage)

	ordered, err := decodeOrderedObject(data)
	if err != nil {
		return err
	}

	sawDeltaUpdate := false
	for _, entry := range ordered {
		switch entry.Key {
		case "deltaUpdate":
			sawDeltaUpdate = true
			if err := d.decodeDeltaOps(entry.Value); err != nil {
				return err
			}
		case "version", "tracks":
			return fmt.Errorf("msf: independent catalog fields are not allowed in a delta catalog")
		case "generatedAt":
			var value int64
			if err := json.Unmarshal(entry.Value, &value); err != nil {
				return err
			}
			d.GeneratedAt = &value
		case "isComplete":
			if err := json.Unmarshal(entry.Value, &d.IsComplete); err != nil {
				return err
			}
		default:
			d.ExtraFields[entry.Key] = cloneRawMessage(entry.Value)
		}
	}
	if !sawDeltaUpdate {
		return fmt.Errorf("msf: delta catalog must include a deltaUpdate array")
	}

	return nil
}

// decodeDeltaOps decodes the draft-01 deltaUpdate array of {op, tracks} objects
// into the grouped slices, recording first-seen block order.
func (d *CatalogDelta) decodeDeltaOps(data []byte) error {
	var ops []map[string]json.RawMessage
	if err := json.Unmarshal(data, &ops); err != nil {
		return fmt.Errorf("msf: deltaUpdate must be an array of operation objects: %w", err)
	}
	for _, opObj := range ops {
		opRaw, ok := opObj["op"]
		if !ok {
			return fmt.Errorf("msf: delta update operation must contain an op field")
		}
		var op string
		if err := json.Unmarshal(opRaw, &op); err != nil {
			return err
		}
		tracksRaw, ok := opObj["tracks"]
		if !ok {
			return fmt.Errorf("msf: delta update operation must contain a tracks field")
		}
		switch deltaOperationKind(op) {
		case deltaOperationAdd:
			if err := json.Unmarshal(tracksRaw, &d.AddTracks); err != nil {
				return err
			}
			d.recordOp(deltaOperationAdd)
		case deltaOperationRemove:
			if err := json.Unmarshal(tracksRaw, &d.RemoveTracks); err != nil {
				return err
			}
			d.recordOp(deltaOperationRemove)
		case deltaOperationClone:
			if err := json.Unmarshal(tracksRaw, &d.CloneTracks); err != nil {
				return err
			}
			d.recordOp(deltaOperationClone)
		default:
			return fmt.Errorf("msf: unknown delta update op %q", op)
		}
	}
	return nil
}

// recordOp appends kind to the declared operation order, ignoring repeats so
// only first-seen block order is preserved.
func (d *CatalogDelta) recordOp(kind deltaOperationKind) {
	if slices.Contains(d.deltaOpOrder, kind) {
		return
	}
	d.deltaOpOrder = append(d.deltaOpOrder, kind)
}

// Clone returns a deep copy of the reference.
func (r TrackRef) Clone() TrackRef {
	clone := r
	clone.ExtraFields = cloneRawMessages(r.ExtraFields)
	return clone
}

// ID returns the resolved identity of the referenced track.
func (r TrackRef) ID(defaultNamespace string) TrackID {
	return TrackID{
		Namespace: r.effectiveNamespace(defaultNamespace),
		Name:      r.Name,
	}
}

// effectiveNamespace resolves Namespace against the delta or catalog default namespace.
func (r TrackRef) effectiveNamespace(defaultNamespace string) string {
	if r.Namespace != "" {
		return r.Namespace
	}
	if defaultNamespace != "" {
		return defaultNamespace
	}
	return inheritedNamespaceSentinel
}

// Validate checks whether the reference is valid for a removeTracks entry.
func (r TrackRef) Validate(path string) []string {
	var problems []string
	prefix := ""
	if path != "" {
		prefix = path + ": "
	}
	if r.Name == "" {
		problems = append(problems, prefix+"name is required")
	}
	if len(r.ExtraFields) > 0 {
		problems = append(problems, prefix+"remove track entries may contain only name and optional namespace")
	}
	return problems
}

// MarshalJSON encodes the reference as a JSON object.
func (r TrackRef) MarshalJSON() ([]byte, error) {
	obj := make(map[string]any, len(r.ExtraFields)+2)
	for key, raw := range r.ExtraFields {
		obj[key] = cloneRawMessage(raw)
	}
	if r.Namespace != "" {
		obj["namespace"] = r.Namespace
	}
	if r.Name != "" {
		obj["name"] = r.Name
	}
	return json.Marshal(obj)
}

// UnmarshalJSON decodes the removeTracks JSON shape.
func (r *TrackRef) UnmarshalJSON(data []byte) error {
	*r = TrackRef{}
	r.ExtraFields = make(map[string]json.RawMessage)

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	for key, value := range raw {
		switch key {
		case "namespace":
			if err := json.Unmarshal(value, &r.Namespace); err != nil {
				return err
			}
		case "name":
			if err := json.Unmarshal(value, &r.Name); err != nil {
				return err
			}
		default:
			r.ExtraFields[key] = cloneRawMessage(value)
		}
	}
	return nil
}

// Clone returns a deep copy of the clone operation.
func (c TrackClone) Clone() TrackClone {
	return TrackClone{
		Track:           c.Track.Clone(),
		ParentName:      c.ParentName,
		ParentNamespace: c.ParentNamespace,
	}
}

// Validate checks whether the cloneTracks entry is valid.
func (c TrackClone) Validate(path string) []string {
	var problems []string
	prefix := ""
	if path != "" {
		prefix = path + ": "
	}
	if c.Name == "" {
		problems = append(problems, prefix+"name is required")
	}
	if c.ParentName == "" {
		problems = append(problems, prefix+"parentName is required for clone tracks")
	}
	return problems
}

// parentEffectiveNamespace resolves the namespace of the parent track being
// cloned. ParentNamespace takes precedence when set; otherwise the catalog
// default namespace (and finally the inherited sentinel) is used.
func (c TrackClone) parentEffectiveNamespace(defaultNamespace string) string {
	if c.ParentNamespace != "" {
		return c.ParentNamespace
	}
	if defaultNamespace != "" {
		return defaultNamespace
	}
	return inheritedNamespaceSentinel
}

// MarshalJSON encodes the clone entry as a JSON object with parentName and
// optional parentNamespace.
func (c TrackClone) MarshalJSON() ([]byte, error) {
	obj := c.Track.marshalObject()
	if c.ParentName != "" {
		obj["parentName"] = c.ParentName
	}
	if c.ParentNamespace != "" {
		obj["parentNamespace"] = c.ParentNamespace
	}
	return json.Marshal(obj)
}

// UnmarshalJSON decodes a cloneTracks entry.
func (c *TrackClone) UnmarshalJSON(data []byte) error {
	*c = TrackClone{}

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	trackRaw := make(map[string]json.RawMessage, len(raw))
	for key, value := range raw {
		switch key {
		case "parentName":
			if err := json.Unmarshal(value, &c.ParentName); err != nil {
				return err
			}
		case "parentNamespace":
			if err := json.Unmarshal(value, &c.ParentNamespace); err != nil {
				return err
			}
		default:
			trackRaw[key] = value
		}
	}
	return c.Track.unmarshalObject(trackRaw)
}

// cloneTrackRefs returns a deep copy of a TrackRef slice.
func cloneTrackRefs(in []TrackRef) []TrackRef {
	if in == nil {
		return nil
	}
	out := make([]TrackRef, len(in))
	for i, track := range in {
		out[i] = track.Clone()
	}
	return out
}

// cloneTrackClones returns a deep copy of a TrackClone slice.
func cloneTrackClones(in []TrackClone) []TrackClone {
	if in == nil {
		return nil
	}
	out := make([]TrackClone, len(in))
	for i, track := range in {
		out[i] = track.Clone()
	}
	return out
}

File: moq-web/src/msf/catalog_delta.ts
import {
	asRecord,
	type Catalog,
	cloneCatalog,
	cloneTrack,
	decodeText,
	parseTrack,
	type Track,
	trackId,
	validateCatalog,
	validateTrack,
	ValidationError,
	zodSchemaError,
} from "./catalog.ts";
import { z } from "zod";
import { resolveInitData } from "./init_data.ts";

/** Discriminator for the three delta operations. */
export type DeltaOperationKind = "addTracks" | "removeTracks" | "cloneTracks";

/** Identifies a track for removal in a delta update. */
export interface TrackRef {
	namespace?: string;
	name?: string;
	extraFields?: Record<string, unknown>;
}

/** A track derived from a parent track via field overrides. */
export interface TrackClone {
	track: Track;
	parentName?: string;
	/** Namespace of the parent track, when it differs from the clone's. */
	parentNamespace?: string;
}

/**
 * An MSF delta update that adds, removes, or clones tracks in an existing
 * {@link Catalog}.
 */
export interface CatalogDelta {
	defaultNamespace?: string;
	generatedAt?: number;
	isComplete?: boolean;
	addTracks: Track[];
	removeTracks: TrackRef[];
	cloneTracks: TrackClone[];
	extraFields?: Record<string, unknown>;
	deltaOpOrder?: DeltaOperationKind[];
}

const trackRefSchema = z.object({
	namespace: z.string().optional(),
	name: z.string().optional(),
}).catchall(z.unknown());

const trackCloneSchema = z.object({
	parentName: z.string().optional(),
	parentNamespace: z.string().optional(),
}).catchall(z.unknown());

// draft-ietf-moq-msf-01: "deltaUpdate" is an array of {op, tracks} objects.
const deltaOperationSchema = z.object({
	op: z.string(),
	tracks: z.array(z.unknown()),
}).catchall(z.unknown());

const catalogDeltaSchema = z.object({
	deltaUpdate: z.array(deltaOperationSchema),
	defaultNamespace: z.string().optional(),
	generatedAt: z.number().optional(),
	isComplete: z.boolean().optional(),
}).catchall(z.unknown());

// Wire op name <-> in-memory operation kind.
const OP_TO_KIND: Record<string, DeltaOperationKind> = {
	add: "addTracks",
	remove: "removeTracks",
	clone: "cloneTracks",
};
const KIND_TO_OP: Record<DeltaOperationKind, string> = {
	addTracks: "add",
	removeTracks: "remove",
	cloneTracks: "clone",
};

function parseTrackRef(value: unknown): TrackRef {
	const rawRecord = asRecord(value, "msf: remove track reference must be a JSON object");
	const parsed = trackRefSchema.safeParse(rawRecord);
	if (!parsed.success) {
		throw zodSchemaError("msf: remove track reference must be a JSON object", parsed.error);
	}
	const raw = parsed.data;
	const extra: Record<string, unknown> = {};
	for (const [key, fieldValue] of Object.entries(raw)) {
		if (key !== "namespace" && key !== "name") {
			extra[key] = fieldValue;
		}
	}
	return {
		namespace: raw.namespace,
		name: raw.name,
		extraFields: Object.keys(extra).length > 0 ? extra : undefined,
	};
}

function parseTrackClone(value: unknown): TrackClone {
	const rawRecord = asRecord(value, "msf: clone track entry must be a JSON object");
	const parsed = trackCloneSchema.safeParse(rawRecord);
	if (!parsed.success) {
		throw zodSchemaError("msf: clone track entry must be a JSON object", parsed.error);
	}
	const raw = parsed.data;
	const { parentName, parentNamespace, ...trackRecord } = raw;
	return {
		track: parseTrack(trackRecord),
		parentName,
		parentNamespace,
	};
}

/**
 * Parse a JSON delta-catalog payload into a {@link CatalogDelta}.
 * @param data - UTF-8 encoded bytes or a JSON string.
 * @throws Error if the payload is invalid or contains independent catalog fields.
 */
export function parseCatalogDelta(data: string | Uint8Array): CatalogDelta {
	const decoded = JSON.parse(decodeText(data));
	const rawRoot = asRecord(decoded, "msf: expected JSON object");
	const parsed = catalogDeltaSchema.safeParse(rawRoot);
	if (!parsed.success) {
		throw zodSchemaError("msf: expected JSON object", parsed.error);
	}
	const root = parsed.data;
	if ("version" in root || "tracks" in root) {
		throw new Error("msf: independent catalog fields are not allowed in a delta catalog");
	}

	// Repeated operations of one kind are legal in draft-01: their tracks are
	// appended, and only the first-seen order of each kind is recorded.
	const deltaOpOrder: DeltaOperationKind[] = [];
	const addTracks: Track[] = [];
	const removeTracks: TrackRef[] = [];
	const cloneTracks: TrackClone[] = [];
	for (const operation of root.deltaUpdate) {
		const kind = OP_TO_KIND[operation.op];
		if (kind === undefined) {
			throw new Error(`msf: unknown delta update op ${JSON.stringify(operation.op)}`);
		}
		if (!deltaOpOrder.includes(kind)) {
			deltaOpOrder.push(kind);
		}
		switch (kind) {
			case "addTracks":
				addTracks.push(...operation.tracks.map(parseTrack));
				break;
			case "removeTracks":
				removeTracks.push(...operation.tracks.map(parseTrackRef));
				break;
			case "cloneTracks":
				cloneTracks.push(...operation.tracks.map(parseTrackClone));
				break;
		}
	}

	const extraFields: Record<string, unknown> = {};
	for (const [key, value] of Object.entries(root)) {
		if (
			key !== "deltaUpdate" &&
			key !== "defaultNamespace" &&
			key !== "generatedAt" &&
			key !== "isComplete"
		) {
			extraFields[key] = value;
		}
	}

	return {
		defaultNamespace: root.defaultNamespace,
		generatedAt: root.generatedAt,
		isComplete: root.isComplete === true,
		addTracks,
		removeTracks,
		cloneTracks,
		extraFields: Object.keys(extraFields).length > 0 ? extraFields : undefined,
		deltaOpOrder,
	};
}

/**
 * Validate a {@link CatalogDelta} according to MSF rules.
 * @throws {@link ValidationError} if any problems are found.
 */
export function validateCatalogDelta(delta: CatalogDelta): void {
	const problems: string[] = [];
	if (
		delta.addTracks.length === 0 &&
		delta.removeTracks.length === 0 &&
		delta.cloneTracks.length === 0
	) {
		problems.push("delta catalog must contain addTracks, removeTracks, or cloneTracks");
	}
	for (let i = 0; i < delta.addTracks.length; i++) {
		problems.push(...validateTrack(delta.addTracks[i]!, `addTracks[${i}]`));
	}
	for (let i = 0; i < delta.removeTracks.length; i++) {
		const ref = delta.removeTracks[i]!;
		if (!ref.name) {
			problems.push(`removeTracks[${i}]: name is required`);
		}
		if (ref.extraFields && Object.keys(ref.extraFields).length > 0) {
			problems.push(
				`removeTracks[${i}]: remove track entries may contain only name and optional namespace`,
			);
		}
	}
	for (let i = 0; i < delta.cloneTracks.length; i++) {
		const clone = delta.cloneTracks[i]!;
		if (!clone.track.name) {
			problems.push(`cloneTracks[${i}]: name is required`);
		}
		if (!clone.parentName) {
			problems.push(`cloneTracks[${i}]: parentName is required for clone tracks`);
		}
	}
	if (problems.length > 0) {
		throw new ValidationError(problems);
	}
}

function applyTrackOverrides(base: Track, override: Track): Track {
	const next: Track = cloneTrack(base);
	for (const [key, value] of Object.entries(override)) {
		if (key === "extraFields") {
			continue;
		}
		if (value !== undefined) {
			if (key === "depends" && Array.isArray(value)) {
				// Clone array-typed depends so the override and resulting track do not share the same instance.
				(next as Record<string, unknown>)[key] = [...value];
			} else {
				(next as Record<string, unknown>)[key] = value;
			}
		}
	}
	// initData is resolved from initRef, so an overridden initRef invalidates
	// the payload inherited from the parent; resolveInitData refills it.
	if (override.initRef !== undefined && override.initData === undefined) {
		delete next.initData;
	}
	if (override.extraFields) {
		next.extraFields = {
			...(next.extraFields ?? {}),
			...override.extraFields,
		};
	}
	return next;
}

function hasInheritedNamespaceTracks(catalog: Catalog): boolean {
	return catalog.tracks.some((track) => !track.namespace);
}

/**
 * Apply a validated delta to a base catalog and return the resulting catalog.
 * @param baseCatalog - The current catalog state.
 * @param deltaCatalog - The delta to apply.
 * @throws Error if the delta cannot be safely applied.
 */
export function applyCatalogDelta(baseCatalog: Catalog, deltaCatalog: CatalogDelta): Catalog {
	validateCatalog(baseCatalog);
	validateCatalogDelta(deltaCatalog);

	const result = cloneCatalog(baseCatalog);
	if (
		deltaCatalog.defaultNamespace !== undefined &&
		deltaCatalog.defaultNamespace !== "" &&
		deltaCatalog.defaultNamespace !== result.defaultNamespace
	) {
		if (hasInheritedNamespaceTracks(result)) {
			throw new Error(
				"msf: cannot change default namespace when catalog contains tracks that inherit it",
			);
		}
		result.defaultNamespace = deltaCatalog.defaultNamespace;
	}
	if (deltaCatalog.generatedAt !== undefined) {
		result.generatedAt = deltaCatalog.generatedAt;
	}
	if (deltaCatalog.isComplete) {
		result.isComplete = true;
	}
	const mergedExtraFields = {
		...(result.extraFields ?? {}),
		...(deltaCatalog.extraFields ?? {}),
	};
	result.extraFields = Object.keys(mergedExtraFields).length > 0 ? mergedExtraFields : undefined;

	const order = deltaCatalog.deltaOpOrder && deltaCatalog.deltaOpOrder.length > 0
		? deltaCatalog.deltaOpOrder
		: [
			deltaCatalog.addTracks.length > 0 ? "addTracks" : undefined,
			deltaCatalog.removeTracks.length > 0 ? "removeTracks" : undefined,
			deltaCatalog.cloneTracks.length > 0 ? "cloneTracks" : undefined,
		].filter((v): v is DeltaOperationKind => v !== undefined);

	for (const op of order) {
		switch (op) {
			case "addTracks":
				for (const track of deltaCatalog.addTracks) {
					const id = trackId(track, result.defaultNamespace);
					if (
						result.tracks.some((candidate) =>
							trackId(candidate, result.defaultNamespace) === id
						)
					) {
						throw new Error(`msf: cannot add duplicate track ${JSON.stringify(id)}`);
					}
					result.tracks.push(cloneTrack(track));
				}
				break;
			case "removeTracks":
				for (const ref of deltaCatalog.removeTracks) {
					const id = trackId(ref, result.defaultNamespace);
					const index = result.tracks.findIndex(
						(track) => trackId(track, result.defaultNamespace) === id,
					);
					if (index < 0) {
						throw new Error(`msf: cannot remove unknown track ${JSON.stringify(id)}`);
					}
					result.tracks.splice(index, 1);
				}
				break;
			case "cloneTracks":
				for (const clone of deltaCatalog.cloneTracks) {
					// As in the Go package: the parent lives in parentNamespace when
					// given, else the catalog default -- not the clone's own namespace.
					const parentId = trackId(
						{ namespace: clone.parentNamespace, name: clone.parentName },
						result.defaultNamespace,
					);
					const parent = result.tracks.find(
						(track) => trackId(track, result.defaultNamespace) === parentId,
					);
					if (!parent) {
						throw new Error(
							`msf: cannot clone unknown parent track ${JSON.stringify(parentId)}`,
						);
					}
					const derived = applyTrackOverrides(parent, clone.track);
					if (!derived.name) {
						throw new Error(
							`msf: cloned track derived from ${
								JSON.stringify(parentId)
							} is missing name`,
						);
					}
					const id = trackId(derived, result.defaultNamespace);
					if (
						result.tracks.some((track) =>
							trackId(track, result.defaultNamespace) === id
						)
					) {
						throw new Error(
							`msf: cannot clone into duplicate track ${JSON.stringify(id)}`,
						);
					}
					result.tracks.push(derived);
				}
		}
	}

	resolveInitData(result);
	validateCatalog(result);
	return result;
}

/**
 * Serialize a {@link CatalogDelta} to its draft-ietf-moq-msf-01 JSON form: a
 * `deltaUpdate` array of `{op, tracks}` objects, in {@link CatalogDelta.deltaOpOrder}
 * when it is set.
 *
 * A delta has no Initialization Data List, so a track can only reference an
 * entry of the base catalog through `initRef`; its resolved `initData` is not
 * written.
 * @throws Error if a track carries `initData` without an `initRef`, since
 * that payload cannot be represented in a delta.
 */
export function stringifyCatalogDelta(delta: CatalogDelta): string {
	const obj: Record<string, unknown> = {
		...(delta.extraFields ?? {}),
	};
	if (delta.defaultNamespace !== undefined) {
		obj.defaultNamespace = delta.defaultNamespace;
	}
	if (delta.generatedAt !== undefined) {
		obj.generatedAt = delta.generatedAt;
	}
	if (delta.isComplete) {
		obj.isComplete = true;
	}

	const trackWire = (track: Track, where: string): Record<string, unknown> => {
		const { extraFields, initData, ...rest } = track;
		if (initData !== undefined && !rest.initRef) {
			throw new Error(
				`msf: ${where} carries initData, which a delta cannot represent; ` +
					"reference an initDataList entry of the base catalog with initRef, " +
					"or publish an independent catalog",
			);
		}
		return { ...(extraFields ?? {}), ...rest };
	};

	const order: DeltaOperationKind[] = delta.deltaOpOrder && delta.deltaOpOrder.length > 0
		? delta.deltaOpOrder
		: (["addTracks", "removeTracks", "cloneTracks"] as const).filter((kind) =>
			delta[kind].length > 0
		);
	obj.deltaUpdate = order.map((kind) => {
		let tracks: Record<string, unknown>[];
		switch (kind) {
			case "addTracks":
				tracks = delta.addTracks.map((track, i) => trackWire(track, `addTracks[${i}]`));
				break;
			case "removeTracks":
				tracks = delta.removeTracks.map((ref) => ({
					...(ref.extraFields ?? {}),
					name: ref.name,
					...(ref.namespace ? { namespace: ref.namespace } : {}),
				}));
				break;
			case "cloneTracks":
				tracks = delta.cloneTracks.map((clone, i) => ({
					...trackWire(clone.track, `cloneTracks[${i}]`),
					parentName: clone.parentName,
					...(clone.parentNamespace ? { parentNamespace: clone.parentNamespace } : {}),
				}));
				break;
		}
		return { op: KIND_TO_OP[kind], tracks };
	});
	return JSON.stringify(obj);
}
